tests/delete/tests.py

```diff
@@ -800,3 +838,189 @@ def test_fast_delete_full_match(self):
+        # Default parent does not exists
+        # Default parent deleted after parent
+        default_parent = ParentModel.objects.create(pk=1)
+        parent = ParentModel.objects.create(pk=2)
+        child1 = DBDefaultsFK.objects.create(parent_model=parent)
+        parent.delete()
+        with self.assertRaises(IntegrityError):
+            default_parent.delete()
+            child1.refresh_from_db()
+
+
+class DatabaseLevelOnDeleteQueryAssertionTests(TestCase):
+    def test_queries_on_nested_cascade(self):
+        parent = ParentModel.objects.create()
+
+        for i in range(3):
+            ChildModel.objects.create(parent_model=parent)
+
+        for child in ChildModel.objects.all():
+            for i in range(3):
+                GrandChild.objects.create(child_model=child)
+
+        # one is the deletion
+        with self.assertNumQueries(1):
+            parent.delete()
+
+    def test_queries_on_nested_set_null(self):
+        parent = ParentModel.objects.create()
+
+        for i in range(3):
+            ChildFKSetNull.objects.create(parent_model=parent)
+
+        for child_fk_set_null in ChildFKSetNull.objects.all():
+            for i in range(3):
+                AnotherGrandChildFKSetNull.objects.create(
+                    child_fk_set_null=child_fk_set_null
+                )
+
+        # one is the deletion
+        with self.assertNumQueries(1):
+            parent.delete()
+
+    def test_queries_on_nested_set_null_cascade(self):
+        parent = ParentModel.objects.create()
+
+        for i in range(3):
+            ChildModel.objects.create(parent_model=parent)
+
+        for child in ChildModel.objects.all():
+            for i in range(3):
+                GrandChildFKSetNull.objects.create(child_model=child)
+
+        # one is the deletion
+        with self.assertNumQueries(1):
+            parent.delete()
+
+    def test_queries_on_inherited_model(self):
+        gp = ParentModel.objects.create()
+        parent = NormalParent.objects.create(grandparent_ptr=gp)
+        diamond_parent = DiamondParent.objects.create(gp_ptr=gp)
+
+        dc = DiamondChild.objects.create(
+            parent_ptr=parent, diamondparent_ptr=diamond_parent
+        )
+
+        with self.assertNumQueries(1):
+            gp.delete()
+
+        self.assertIs(NormalParent.objects.filter(pk=parent.pk).exists(), False)
+        self.assertIs(
+            DiamondParent.objects.filter(pk=diamond_parent.pk).exists(), False
+        )
+        self.assertIs(DiamondChild.objects.filter(pk=dc.pk).exists(), False)
+
+    def test_restrict_on_inherited_model(self):
+        gp = ParentModel.objects.create()
+        child = DBLevelChild.objects.create(grandparent_ptr=gp)
+
+        with transaction.atomic():
+            with self.assertRaises(IntegrityError):
+                gp.delete()
+
+        self.assertIs(DBLevelChild.objects.filter(pk=child.pk).exists(), True)
```

Comment:
IMO most of these tests are unnecessary, we don't do anything specific here so testing all different models configuration are overkill. We don't want to test database behavior.